### ISSUE TITLE
false positive for add reference here

### ISSUE THREAD

**Issue by Heinenen** [rust-lang/rust-analyzer#12575]
When browsing through the r-a code rust-analyzer gave me an error despite everything compiling correctly, saying I need to "add a reference here".
The r-a error shows on `, position)` in both lines:
https://github.com/rust-lang/rust-analyzer/blob/12dd81092e37df28b7a3591cae9675e668927198/crates/rust-analyzer/src/handlers.rs#L1731-L1732

I managed to isolate the behavior (the code compiles but r-a show error on lines 4, 5):
``` rust
fn main() {
    let x = vec![Enum::First(Struct { x:5 }), Enum::Second(Struct {x:1})];
    let result: Vec<u32> = x.iter().filter_map(|it| match it {
        Enum::First(content) => Some(unpack(content)),
        Enum::Second(content) => Some(unpack(content)),
    }).collect();
    println!("{:?}", result);
}

struct Struct {
    x: u32,
}
enum Enum {
    First(Struct),
    Second(Struct),
}

fn unpack(param: &Struct) -> u32 {
    param.x
}
```
Using 
``` rust 
unpack(Struct { x:3 });
```
or 
``` rust
match x[0] {
    Enum::First(content) => unpack(content),
    Enum::Second(content) => unpack(content),
};
```
causes the program to not compile and r-a showing an error, but this time in accordance with the compiler.

</br>
**rust-analyzer version**: db2a7087b 2021-12-13 stable

**rustc version**: 1.61.0 (fe5b13d68 2022-05-18)

**Comment by flodiebold** [comment]
You're using a very old rust-analyzer version. If you're using VSCode, try reinstalling the extension, otherwise I suggest updating in whatever way you installed it :smile: 
This is fixed in current versions.

**Comment by lnicola** [comment]
If you're on VSCodium, you can get a VSIX from the GitHub releases page.

**Comment by Heinenen** [comment]
Probably should've read the version, not just copy-pasted it 😄 
With the new version it now works.
